File: friday/session.py
"""Transport-agnostic Gemini Live session.

A transport pushes 16 kHz mono s16le PCM in via `send_audio()` and consumes
`Event`s from `events()`. Tool calls are dispatched to the registry in the
background so audio keeps streaming (Gemini 3.8 Live calls are NON_BLOCKING
by default).
"""
from __future__ import annotations

import asyncio
import logging
from dataclasses import dataclass
from typing import AsyncIterator, Literal

from google import genai
from google.genai import types

from .config import settings
from .tools import registry

log = logging.getLogger(__name__)


@dataclass
class Event:
    kind: Literal["audio", "interrupted", "turn_complete", "user_text", "bot_text", "tool_call", "tool_result", "closed"]
    data: bytes | str | dict | None = None


class GeminiSession:
    def __init__(self) -> None:
        self._client = genai.Client(api_key=settings.api_key)
        self._events: asyncio.Queue[Event] = asyncio.Queue()
        self._session = None
        self._tasks: set[asyncio.Task] = set()
        self._closed = asyncio.Event()

    def _config(self) -> dict:
        return {
            "response_modalities": ["AUDIO"],
            "system_instruction": settings.system_prompt,
            "input_audio_transcription": {},
            "output_audio_transcription": {},
            "speech_config": {
                "voice_config": {"prebuilt_voice_config": {"voice_name": settings.voice}}
            },
            "tools": [{"function_declarations": registry.declarations()}],
        }

    async def __aenter__(self) -> "GeminiSession":
        self._cm = self._client.aio.live.connect(model=settings.model, config=self._config())
        self._session = await self._cm.__aenter__()
        self._spawn(self._receive_loop())
        log.info("gemini session open (%s, %d tools)", settings.model, len(registry.declarations()))
        return self

    async def __aexit__(self, *exc) -> None:
        await self.close()

    def _spawn(self, coro) -> None:
        t = asyncio.create_task(coro)
        self._tasks.add(t)
        t.add_done_callback(self._tasks.discard)

    # ---- inbound ---------------------------------------------------------
    async def send_audio(self, pcm16k: bytes) -> None:
        if self._session and not self._closed.is_set():
            await self._session.send_realtime_input(
                audio=types.Blob(data=pcm16k, mime_type=f"audio/pcm;rate={settings.input_rate}")
            )

    async def send_text(self, text: str) -> None:
        if self._session:
            await self._session.send_client_content(turns={"role": "user", "parts": [{"text": text}]})

    # ---- outbound --------------------------------------------------------
    async def events(self) -> AsyncIterator[Event]:
        while True:
            ev = await self._events.get()
            yield ev
            if ev.kind == "closed":
                return

    async def _receive_loop(self) -> None:
        try:
            async for msg in self._session.receive():
                if msg.tool_call:
                    for fc in msg.tool_call.function_calls:
                        self._spawn(self._run_tool(fc))
                sc = msg.server_content
                if not sc:
                    continue
                if sc.interrupted:
                    await self._events.put(Event("interrupted"))
                if sc.input_transcription and sc.input_transcription.text:
                    await self._events.put(Event("user_text", sc.input_transcription.text))
                if sc.output_transcription and sc.output_transcription.text:
                    await self._events.put(Event("bot_text", sc.output_transcription.text))
                if sc.model_turn:
                    for part in sc.model_turn.parts:
                        if part.inline_data and part.inline_data.data:
                            await self._events.put(Event("audio", part.inline_data.data))
                if sc.turn_complete:
                    await self._events.put(Event("turn_complete"))
        except Exception:  # noqa: BLE001
            log.exception("receive loop ended")
        finally:
            await self._events.put(Event("closed"))

    async def _run_tool(self, fc) -> None:
        args = dict(fc.args or {})
        await self._events.put(Event("tool_call", {"name": fc.name, "args": args}))
        result, scheduling = await registry.call(fc.name, args)
        await self._events.put(Event("tool_result", {"name": fc.name, "result": result}))
        payload = {**result, "scheduling": scheduling}
        try:
            await self._session.send_tool_response(
                function_responses=[types.FunctionResponse(id=fc.id, name=fc.name, response=payload)]
            )
        except Exception:  # noqa: BLE001
            log.exception("failed to send tool response for %s", fc.name)

    async def close(self) -> None:
        if self._closed.is_set():
            return
        self._closed.set()
        for t in list(self._tasks):
            t.cancel()
        try:
            await self._cm.__aexit__(None, None, None)
        except Exception:  # noqa: BLE001
            pass

File: src/session.ts
/**
 * Transport-agnostic Gemini Live session.
 * Push 16 kHz mono s16le PCM in with sendAudio(); receive Events via onEvent.
 * Tool calls run in the background so audio keeps streaming.
 */
import { GoogleGenAI, Modality, type LiveServerMessage, type Session } from "@google/genai";
import { settings } from "./config.js";
import { callTool, declarations } from "./tools/index.js";
import { END_CONVERSATION } from "./tools/builtin.js";

export type Event =
  | { kind: "audio"; data: Buffer }
  | { kind: "interrupted" }
  | { kind: "turn_complete" }
  | { kind: "user_text"; data: string }
  | { kind: "bot_text"; data: string }
  | { kind: "tool_call"; data: { name: string; args: unknown } }
  | { kind: "tool_result"; data: { name: string; result: unknown } }
  | { kind: "closed"; data?: string };

export class GeminiSession {
  private session?: Session;
  private closed = false;
  /** Set when the model called end_conversation; we close after its turn finishes. */
  private endRequested?: string;
  private idleTimer?: NodeJS.Timeout;
  private toolsInFlight = 0;

  constructor(private readonly onEvent: (e: Event) => void) {}

  async open(): Promise<void> {
    const ai = new GoogleGenAI({ apiKey: settings.apiKey });
    const decls = declarations();
    this.session = await ai.live.connect({
      model: settings.model,
      config: {
        responseModalities: [Modality.AUDIO],
        systemInstruction: settings.systemPrompt,
        inputAudioTranscription: {},
        outputAudioTranscription: {},
        speechConfig: { voiceConfig: { prebuiltVoiceConfig: { voiceName: settings.voice } } },
        tools: [{ functionDeclarations: decls }],
      },
      callbacks: {
        onmessage: (m) => this.handle(m),
        onerror: (e) => console.error("gemini error", e.message),
        onclose: (e) => this.emitClosed(e.reason),
      },
    });
    console.log(`gemini session open (${settings.model}, ${decls.length} tools)`);
  }

  async sendAudio(pcm16k: Buffer): Promise<void> {
    if (this.closed) return;
    this.session?.sendRealtimeInput({
      audio: { data: pcm16k.toString("base64"), mimeType: `audio/pcm;rate=${settings.inputRate}` },
    });
  }

  sendText(text: string): void {
    this.session?.sendClientContent({ turns: [{ role: "user", parts: [{ text }] }] });
  }

  private handle(m: LiveServerMessage): void {
    for (const fc of m.toolCall?.functionCalls ?? []) void this.runTool(fc.id, fc.name, fc.args);
    const sc = m.serverContent;
    if (!sc) return;
    // Gemini flags its own turn as interrupted when the end_conversation tool response
    // arrives; forwarding that would make clients cut Friday's final words.
    if (sc.interrupted && !this.endRequested) this.onEvent({ kind: "interrupted" });
    if (sc.inputTranscription?.text) {
      this.clearIdle(); // the user is talking again
      this.onEvent({ kind: "user_text", data: sc.inputTranscription.text });
    }
    if (sc.outputTranscription?.text) this.onEvent({ kind: "bot_text", data: sc.outputTranscription.text });
    for (const p of sc.modelTurn?.parts ?? []) {
      if (p.inlineData?.data) this.onEvent({ kind: "audio", data: Buffer.from(p.inlineData.data, "base64") });
    }
    if (sc.turnComplete) {
      this.onEvent({ kind: "turn_complete" });
      if (this.endRequested) this.finish(`ended: ${this.endRequested}`);
      else this.armIdle();
    }
  }

  /** Close once the user has been silent for idleTimeoutMs after a turn, unless a tool is still pending. */
  private armIdle(): void {
    this.clearIdle();
    if (!settings.idleTimeoutMs || this.toolsInFlight > 0) return;
    this.idleTimer = setTimeout(() => this.finish("ended: no follow-up"), settings.idleTimeoutMs);
  }

  private clearIdle(): void {
    if (this.idleTimer) clearTimeout(this.idleTimer);
    this.idleTimer = undefined;
  }

  private finish(reason: string): void {
    if (this.closed) return;
    console.log(`gemini session ${reason}`);
    this.session?.close();
    this.emitClosed(reason);
  }

  private async runTool(id?: string, name?: string, args?: Record<string, unknown>): Promise<void> {
    if (!name) return;
    this.onEvent({ kind: "tool_call", data: { name, args } });
    this.toolsInFlight++;
    this.clearIdle();
    const { result, scheduling } = await callTool(name, args);
    this.toolsInFlight--;
    this.onEvent({ kind: "tool_result", data: { name, result } });
    if (name === END_CONVERSATION) this.endRequested = String(args?.reason ?? "done");
    this.session?.sendToolResponse({ functionResponses: [{ id, name, response: { ...result, scheduling } }] });
  }

  private emitClosed(reason?: string): void {
    if (this.closed) return;
    this.closed = true;
    this.clearIdle();
    this.onEvent({ kind: "closed", data: reason });
  }

  close(): void {
    if (this.closed) return;
    this.session?.close();
    this.emitClosed("client closed");
  }
}
